artifactory/commands/terraform/terraformpublish.go

```diff
@@ -0,0 +1,365 @@
+}
+
+func (tpc *TerraformPublishCommand) publish() error {
+	log.Debug("Deploying terraform module.")
+	success, failed, err := tpc.terraformPublish()
+	if err != nil {
+		return err
+	}
+	tpc.result.SetSuccessCount(success)
+	tpc.result.SetFailCount(failed)
+	return nil
+}
+
+func (tpa *TerraformPublishCommandArgs) extractTerraformPublishOptionsFromArgs(args []string) (err error) {
+	// Extract namespace information from the args.
+	var flagIndex, valueIndex int
+	flagIndex, valueIndex, tpa.namespace, err = coreutils.FindFlag("--namespace", args)
+	if err != nil {
+		return
+	}
+	coreutils.RemoveFlagFromCommand(&args, flagIndex, valueIndex)
+	// Extract provider information from the args.
+	flagIndex, valueIndex, tpa.provider, err = coreutils.FindFlag("--provider", args)
+	if err != nil {
+		return
+	}
+	coreutils.RemoveFlagFromCommand(&args, flagIndex, valueIndex)
+	// Extract tag information from the args.
+	flagIndex, valueIndex, tpa.tag, err = coreutils.FindFlag("--tag", args)
+	if err != nil {
+		return
+	}
+	coreutils.RemoveFlagFromCommand(&args, flagIndex, valueIndex)
+	// Extract exclusions information from the args.
+	flagIndex, valueIndex, exclusionsString, err := coreutils.FindFlag("--exclusions", args)
+	if err != nil {
+		return
+	}
+	for _, singleValue := range strings.Split(exclusionsString, ";") {
+		tpa.exclusions = append(tpa.exclusions, singleValue)
+	}
+	coreutils.RemoveFlagFromCommand(&args, flagIndex, valueIndex)
+	if len(args) != 0 {
+		err = errorutils.CheckErrorf("Unknown flag:" + strings.Split(args[0], "=")[0] + ". for a terraform publish command please provide --namespace, --provider, --tag and optionally --exclusions.")
+	}
+	return
+}
+
+func (tpc *TerraformPublishCommand) terraformPublish() (int, int, error) {
+	uploadSummary := clientservicesutils.NewResult(cliutils.Threads)
+	producerConsumer := parallel.NewRunner(cliutils.Threads, 20000, false)
+	errorsQueue := clientutils.NewErrorsQueue(1)
+
+	tpc.prepareTerraformPublishTasks(producerConsumer, errorsQueue, uploadSummary)
+	totalUploaded, totalFailed := tpc.performTerraformPublishTasks(producerConsumer, uploadSummary)
+	e := errorsQueue.GetError()
+	if e != nil {
+		return 0, 0, e
+	}
+	return totalUploaded, totalFailed, nil
+}
+
+func (tpc *TerraformPublishCommand) prepareTerraformPublishTasks(producer parallel.Runner, errorsQueue *clientutils.ErrorsQueue, uploadSummary *clientservicesutils.Result) {
+	go func() {
+		defer producer.Done()
+		pwd, err := os.Getwd()
+		if err != nil {
+			log.Error(err)
+			errorsQueue.AddError(err)
+		}
+		// Walk and upload directories which contain '.tf' files.
+		err = tpc.walkDirAndUploadTerraformModules(pwd, producer, errorsQueue, uploadSummary, addTaskWithError)
+		if err != nil && err != io.EOF {
+			log.Error(err)
+			errorsQueue.AddError(err)
+		}
+	}()
+}
+
+// ProduceTaskFunk Provided as a parameter to 'walkDirAndUploadTerraformModules' function for testing reasons.
```

Comment:
ProduceTaskFunk Provided as a parameter... --> ProduceTaskFunk is provided as an argument...
reasons --> purposes